src/python/zquantum/core/wip/circuit/_circuit.py

```diff
@@ -0,0 +1,148 @@
+import json
+from .gates import Gate
+from ...utils import SCHEMA_VERSION
+from typing import List, Dict, Union, TextIO
+
+
+class Circuit(object):
+    """Base class for quantum circuits.
+    
+    Attributes:
+        gates (List[Gate]): The sequence of gates to be applied
+        qubits (tuple(int)): The set of qubits that the circuit acts on
+        symbolic_params (set(sympy.Symbol)): A set of the parameter names used in the circuit. If the circuit
+            does not contain parameterized gates, this value is the empty set.
+    """
+
+    def __init__(self, gates: List[Gate] = None):
+        """Initialize a quantum circuit 
+
+        Args:
+            gates (List[Gate]): See class definition
+            qubits (tuple(int)): See class definition
+        """
+        if gates is None:
+            self.gates = []
+        else:
+            self.gates = gates
+
+    @property
+    def qubits(self):
+        """ The qubits that are used by the gates in the circuit
+
+        Returns:
+            tuple(int)
+        """
+        qubits = {
+            qubit
+            for gate in self.gates
+            for qubit in gate.qubits
+        }
+        return tuple(sorted(qubits))
+
+    @property
+    def symbolic_params(self):
+        """ The set of symbolic parameters used in the circuit
+
+        Returns:
+            set: set of all the sympy symbols used as params of gates in the circuit.
+        """
+        symbolic_params = []
+        for gate in self.gates:
+            symbolic_params_per_gate = gate.symbolic_params
+            symbolic_params += symbolic_params_per_gate
+
+        return set(symbolic_params)
```

Comment:
@mstechly I forget where we landed on this. Do we want this to be a list instead of a set so that the order is assured?